File: thesis/algorithms/javascript/p.js
const fs = require('fs');

function mergeSort(arr) {
    if (arr.length <= 1) return arr;
    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));
    return merge(left, right);
}

function merge(left, right) {
    const result = [];
    let i = 0, j = 0;
    while (i < left.length && j < right.length) {
        result.push(left[i] <= right[j] ? left[i++] : right[j++]);
    }
    return result.concat(left.slice(i)).concat(right.slice(j));
}

function main() {
    const size = process.argv[2] || 'small';
    const path = `../data/${size}/merge_sort_${size}.json`;
    if (!fs.existsSync(path)) {
        console.error('Arquivo não encontrado.');
        return;
    }
    const arr = JSON.parse(fs.readFileSync(path));
    mergeSort(arr);
    console.log(`Ordenado ${arr.length} elementos (${size})`);
}

main();

File: thesis/algorithms/python/p.py
import json
import sys

def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)

def merge(left, right):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

def main(size='small'):
    with open(f'datasets/{size}/merge_sort.json', 'r') as f:
        arr = json.load(f)
    sorted_arr = merge_sort(arr)
    print(f'Sorted {size} dataset with {len(arr)} elements.')

if __name__ == '__main__':
    size = sys.argv[1] if len(sys.argv) > 1 else 'small'
    main(size)
